sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -193,38 +195,68 @@ case class InMemoryTableScanExec(
 
   private val inMemoryPartitionPruningEnabled = sqlContext.conf.inMemoryPartitionPruning
 
+  private def doFilterCachedBatches(
+      rdd: RDD[CachedBatch],
+      partitionStatsSchema: Seq[AttributeReference]): RDD[CachedBatch] = {
+    val schemaIndex = partitionStatsSchema.zipWithIndex
+    rdd.mapPartitionsWithIndex {
+      case (partitionIndex, cachedBatches) =>
+        if (inMemoryPartitionPruningEnabled) {
+          cachedBatches.filter { cachedBatch =>
+            val partitionFilter = newPredicate(
```

Comment:
can this be pulled up out of `usePartitionLevelMetadata` ? seems like you're constructing the predicate per record